### ISSUE TITLE
Add various loss functions to the api

### ISSUE THREAD

**Issue by nsthorat** [tensorflow/tfjs#131]
_From @ManrajGrover on February 22, 2018 19:6_

Currently, the api only supports Softmax Cross Entropy. I feel we can other loss functions to the api. Listing them down here:

1. Absolute Difference
2. Cosine Distance
3. Hinge Loss
4. Huber Loss
5. Log Loss
6. Mean Pairwise Squared Error
7. Mean Squared Error
8. Sigmoid Cross Entropy Loss

I can work on these one by one.

_Copied from original issue: tensorflow/tfjs-core#770_

**Comment by nsthorat** [comment]
Adding @caisq

**Comment by ManrajGrover** [comment]
@nsthorat @caisq Taking this one up, one at a time starting with Mean Squared Error.

**Comment by ManrajGrover** [comment]
@nsthorat With recent discussions (https://github.com/tensorflow/tfjs-core/pull/868#issuecomment-379477113) in mind, should I proceed with implementing this or wait for a while?

**Comment by dsmilkov** [comment]
I think the best way is to go over issues and see what people run into when trying to port their existing models. https://github.com/tensorflow/tfjs/issues/148 is a good example of that. Other folks already started adding some of those. You can comment on that thread to claim one of the ops.

`StridedSlice`, `SpaceToBatchND` and `BatchToSpaceND` come up often.

Thanks Manraj!

**Comment by ManrajGrover** [comment]
Thanks for sharing @dsmilkov. Agree to everything. Since I haven't worked with them, I would need sometime understanding these operations (especially their gradients). If you could share some resources on it, it would be great since it's tough to find gradients for such operations.

Just a side note, I was going through [tensorflow/tfjs-layers](https://github.com/tensorflow/tfjs-layers) and found many of loss functions have already been [implemented](https://github.com/tensorflow/tfjs-layers/blob/master/src/losses.ts). I feel we should move them to tfjs-core and expose them in layers api keeping bundle size sanity, reducing duplication and supporting loss functions in core api.

**Comment by nsthorat** [comment]
+1 I think it make sense to push the loss functions that live in layers into core!

**Comment by ManrajGrover** [comment]
@nsthorat I was thinking of taking this one up. Can I get a go-ahead?

**Comment by nsthorat** [comment]
You have my go ahead :) thanks Manraj

**Comment by ManrajGrover** [comment]
Updating things here:

- [x] Absolute Difference
- [x] Cosine Distance
- [x] Hinge Loss
- [ ] Huber Loss
- [x] Log Loss
- [ ] Mean Pairwise Squared Error
- [x] Mean Squared Error
- [ ] Sigmoid Cross Entropy Loss
- [x] Softmax Cross Entropy Loss
- [ ] Sparse Softmax Cross Entropy Loss

I believe we can make a release and start replacing loss operations implemented in Layers with Core API. I don't think all of the loss functions from Layers API (Keras) can be implemented in Core API to keep things aligned (unless these losses make use of other tensorflow exports). Following is the list for the same:

1. Mean Absolute Percentage Error
2. Mean Squared Logarithmic Error
3. Squared Hinge
4. Categorical Hinge
5. Log cosh
6. Cosine Proximity
7. Poisson
8. Kullback Leibler Divergence

**Comment by ericdnielsen** [comment]
I've been working through some of those differences.  The current approach I'm trying to take is to preserve the Layer's API defaults at the level of any of the objects (Layers, Initializers, etc) in order to preserve serialization compatibility, while migrating internal symbols/functions to match core -- often this means making sure layers always explicitly sets a default.

**Comment by ManrajGrover** [comment]
@ericdnielsen Ah, got it. Thanks for sharing this.

**Comment by ManrajGrover** [comment]
[softmaxCrossEntropy](https://github.com/tensorflow/tfjs-core/blob/master/src/ops/softmax.ts#L107) loss implemented does not support weighted loss. Looking at API params, it actually is called [softmax_cross_entropy_with_logits_v2](https://www.tensorflow.org/api_docs/python/tf/nn/softmax_cross_entropy_with_logits_v2) in Python API. I was planning to rename this function and change loss function such that it supports weighted loss. This might also help in fixing [this](https://github.com/tensorflow/tfjs/issues/188#issuecomment-385290664) issue.

/cc @nsthorat @dsmilkov

**Comment by dsmilkov** [comment]
How about modifying (generalizing) softmaxCrossEntropy in core to align with [tf.losses.softmax_cross_entropy](https://www.tensorflow.org/api_docs/python/tf/losses/softmax_cross_entropy) which does support weights?

**Comment by dsmilkov** [comment]
FYI, `tf.nn.*` is an outdated API and might go away when a major new version of TF gets released.

**Comment by ManrajGrover** [comment]
I was thinking of moving all cross entropy functions (`tf.sigmoidCrossEntropyWithLogits` etc) to a separate class, make `tf.losses.softmaxCrossEntropy` -> `tf.softmaxCrossEntropyWithLogits` and implement `tf.losses.softmaxCrossEntropy` like [`tf.losses.sigmoidCrossEntropy`](https://github.com/tensorflow/tfjs-core/pull/1120)

> FYI, I think tf.nn.* is an outdated API and might go away when a major new version of TF gets released.

I wasn't aware of this. Are they being migrated to `tf.*` namespace?

**Comment by dsmilkov** [comment]
I'm not sure about all methods in `tf.nn.*` but `tf.losses.*` is a much newer namespace than `tf.nn.*` and all loss functions should go in `tf.losses.*`. Assume that losses in `tf.nn.*` may have an outdated API, and the API that matters in the methods in `tf.losses.*`

**Comment by ManrajGrover** [comment]
It makes sense to generalize current implementation of softmax cross entropy then (maybe even move it under `LossOps` class). One issue I see is this change might be a breaking one. Should I go ahead with the change aligning it with Python API?

**Comment by dsmilkov** [comment]
Moving softmaxcrossentropy under LossOps is not a breaking change since it's already [exposed under tf.losses](https://github.com/tensorflow/tfjs-core/blob/master/src/ops/ops.ts#L258). I don't see any breakage in generalizing the API, (just adds more params to the func call). What breakage do you see?

**Comment by ManrajGrover** [comment]
I was thinking it would require removing `dims` param to keep API aligned with Python API (which would break the API).

**Comment by dsmilkov** [comment]
Let's break API since the source of truth should be [tf.losses.softmax_cross_entropy
](https://www.tensorflow.org/api_docs/python/tf/losses/softmax_cross_entropy), not [tf.nn.softmax_cross_entropy_with_logits](https://www.tensorflow.org/api_docs/python/tf/nn/softmax_cross_entropy_with_logits)

**Comment by nsthorat** [comment]
There are just two more to implement here, we can file individual issues for those?
